File: src/main/java/javabot/operations/GetFactoidOperation.java
package javabot.operations;

import java.util.ArrayList;
import java.util.HashSet;
import java.util.List;
import java.util.Set;

import javabot.Action;
import javabot.BotEvent;
import javabot.Javabot;
import javabot.Message;
import javabot.dao.FactoidDao;
import javabot.model.Factoid;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.beans.factory.annotation.Autowired;

public class GetFactoidOperation extends BotOperation {
    private static final Logger log = LoggerFactory.getLogger(GetFactoidOperation.class);
    @Autowired
    private FactoidDao factoidDao;

    public GetFactoidOperation(final Javabot bot) {
        super(bot);
    }

    @Override
    public List<Message> handleMessage(final BotEvent event) {
        final List<Message> messages = new ArrayList<Message>();
        getFactoid(event.getMessage(), event.getSender(), messages, event.getChannel(), event, new HashSet<String>());
        return messages;
    }

    private void getFactoid(final String toFind, final String sender, final List<Message> messages, final String channel, final BotEvent event,
                            final Set<String> backtrack) {
        String message = toFind;
        if(log.isDebugEnabled()) {
            log.debug(sender + " : " + message);
        }
        if (message.endsWith(".") || message.endsWith("?") || message.endsWith("!")) {
            message = message.substring(0, message.length() - 1);
        }
        final String firstWord = message.replaceAll(" .+", "");
        final String dollarOne = message.replaceFirst("[^ ]+ ", "");
        final String key = message;
        if (!factoidDao.hasFactoid(message.toLowerCase()) && factoidDao.hasFactoid(firstWord.toLowerCase() + " $1")) {
            message = firstWord + " $1";
        }
        final Factoid factoid = factoidDao.getFactoid(message.toLowerCase());
        if (factoid != null) {
            message = factoid.getValue();
            message = message.replaceAll("\\$who", sender);
            message = message.replaceAll("\\$1", dollarOne);
            message = processRandomList(message);
            if (message.startsWith("<see>")) {
                if (!backtrack.contains(message)) {
                    backtrack.add(message);
                    getFactoid(message.substring("<see>".length()).trim(), sender, messages, channel, event, backtrack);
                } else {
                    messages.add(new Message(channel, event,
                        "Reference loop detected for factoid '" + message + "'."));
                }
            } else if (message.startsWith("<reply>")) {
                messages.add(new Message(channel, event, message.substring("<reply>".length())));
            } else if (message.startsWith("<action>")) {
                messages.add(new Action(channel, event, message.substring("<action>".length())));
            } else {
                messages.add(new Message(channel, event, sender + ", " + key + " is " + message));
            }
        } else {
            final GuessOperation operation = (GuessOperation) getBot().getOperation(GuessOperation.class.getName());
            final List<Message> guessed = operation.handleMessage(new BotEvent(event.getChannel(),
                    event.getSender(), event.getLogin(), event.getHostname(), "guess " + message));
            final Message guessedMessage = guessed.get(0);
            if (!"No appropriate factoid found.".equals(guessedMessage.getMessage())) {
                messages.addAll(guessed);
            }
        }
        if (messages.isEmpty()) {
            messages.add(new Message(channel, event, sender + ", I have no idea what " + message + " is."));
        }
    }

    protected String processRandomList(final String message) {
        String result = message;
        int index = -1;
        index = result.indexOf("(", index + 1);
        int index2 = result.indexOf(")", index + 1);
        while (index < result.length() && index != -1 && index2 != -1) {
            final String choice = result.substring(index + 1, index2);
            final String[] choices = choice.split("\\|");
            if (choices.length > 1) {
                final int chosen = (int) (Math.random() * choices.length);
                result = result.substring(0, index) + choices[chosen] + result.substring(index2 + 1);
            }
            index = result.indexOf("(", index + 1);
            index2 = result.indexOf(")", index + 1);
        }
        return result;
    }
}

File: src/main/kotlin/javabot/operations/GetFactoidOperation.kt
package javabot.operations

import com.antwerkz.sofia.Sofia
import java.time.LocalDateTime
import java.util.HashSet
import java.util.Locale
import javabot.Action
import javabot.Javabot
import javabot.Message
import javabot.dao.AdminDao
import javabot.dao.FactoidDao
import javabot.model.Factoid
import javax.inject.Inject
import org.slf4j.LoggerFactory

class GetFactoidOperation
@Inject
constructor(bot: Javabot, adminDao: AdminDao, var factoidDao: FactoidDao) :
    BotOperation(bot, adminDao), StandardOperation {

    override fun getPriority(): Int {
        return Integer.MIN_VALUE
    }

    override fun handleMessage(event: Message): List<Message> {
        val responses = arrayListOf<Message>()

        validateTell(responses, event)

        if (responses.isEmpty()) {
            getFactoid(responses, event, HashSet<String>())
        }

        return responses
    }

    private fun getFactoid(
        responses: MutableList<Message>,
        event: Message,
        backtrack: MutableSet<String>,
    ) {
        var message = event.value
        if (message.endsWith(".") || message.endsWith("?") || message.endsWith("!")) {
            message = message.substring(0, message.length - 1)
        }
        val firstWord = message.split(" ")[0]
        val params = message.substring(firstWord.length).trim()
        var factoid = factoidDao.getFactoid(message.lowercase(Locale.getDefault()))
        if (factoid == null) {
            factoid = factoidDao.getParameterizedFactoid(firstWord)
        }

        if (factoid != null) {
            getResponse(responses, event, backtrack, params, factoid)
            factoid.lastUsed = LocalDateTime.now()
            factoid.usage += 1
            factoidDao.save(factoid)
        }
    }

    private fun getResponse(
        responses: MutableList<Message>,
        event: Message,
        backtrack: MutableSet<String>,
        replacedValue: String,
        factoid: Factoid,
    ) {
        val sender = event.user.nick
        val message = factoid.evaluate(event.target, sender, replacedValue)
        if (message.startsWith("<see>")) {
            if (backtrack.contains(message)) {
                responses.add(Message(event, Sofia.factoidLoop(message)))
            } else {
                backtrack.add(message)
                return getFactoid(responses, Message(event, message.substring(5).trim()), backtrack)
            }
        } else if (message.startsWith("<reply>")) {
            responses.add(Message(event, message.substring("<reply>".length)))
        } else if (message.startsWith("<action>")) {
            try {
                responses.add(Action(event, message.substring("<action>".length)))
            } catch (e: Exception) {
                LOG.error(
                    """Exception:  event.target = [${event.target}], event = [${event}], backtrack = [${backtrack}],
                        replacedValue = [${replacedValue}], factoid = [${factoid}]""",
                    e,
                )
            }
        } else {
            responses.add(Message(event, message))
        }
    }

    private fun validateTell(responses: MutableList<Message>, event: Message) {
        if (event.tell) {
            val channel = event.channel
            val targetUser = event.target!!
            if (targetUser.nick.equals(bot.nick, ignoreCase = true)) {
                responses.add(Message(event, Sofia.botSelfTalk()))
            } else {
                if (channel != null) {
                    if (!bot.isOnCommonChannel(targetUser)) {
                        responses.add(
                            Message(event, Sofia.userNotInChannel(targetUser.nick, channel.name))
                        )
                    } else if (
                        event.user.nick == channel.name && !bot.isOnCommonChannel(targetUser)
                    ) {
                        responses.add(Message(event, Sofia.userNoSharedChannels()))
                    }
                }
            }
        }
    }

    companion object {
        private val LOG = LoggerFactory.getLogger(GetFactoidOperation::class.java)
    }
}
